src/mono/wasm/runtime/storage.ts

```diff
@@ -0,0 +1,97 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+import ProductVersion from "consts:productVersion";
+import GitHash from "consts:gitHash";
+
+// adapted from Blazor's WebAssemblyResourceLoader.ts
+async function open(): Promise<Cache | null> {
+    // caches will be undefined if we're running on an insecure origin (secure means https or localhost)
+    if (typeof globalThis.caches === "undefined") {
+        return null;
+    }
+
+    // cache integrity is compromised if the first request has been served over http (except localhost)
+    // in this case, we want to disable caching and integrity validation
+    if (ENVIRONMENT_IS_WEB && globalThis.window.isSecureContext === false) {
+        return null;
+    }
+
+    // Define a separate cache for each base href, so we're isolated from any other
+    // Blazor application running on the same origin. We need this so that we're free
+    // to purge from the cache anything we're not using and don't let it keep growing,
+    // since we don't want to be worst offenders for space usage.
+    const relativeBaseHref = document.baseURI.substring(document.location.origin.length);
+    const cacheName = `dotnet-resources${relativeBaseHref}`;
+
+    try {
+        // There's a Chromium bug we need to be aware of here: the CacheStorage APIs say that when
+        // caches.open(name) returns a promise that succeeds, the value is meant to be a Cache instance.
+        // However, if the browser was launched with a --user-data-dir param that's "too long" in some sense,
+        // then even through the promise resolves as success, the value given is `undefined`.
+        // See https://stackoverflow.com/a/46626574 and https://bugs.chromium.org/p/chromium/issues/detail?id=1054541
+        // If we see this happening, return "null" to mean "proceed without caching".
+        return (await globalThis.caches.open(cacheName)) || null;
+    } catch {
+        // There's no known scenario where we should get an exception here, but considering the
+        // Chromium bug above, let's tolerate it and treat as "proceed without caching".
+        return null;
+    }
+}
+const memoryPrefix = "https://dotnet.generated.invalid/wasm-memory";
+export async function getMemory(configurationHash: number): Promise<ArrayBuffer | undefined> {
+    try {
+        const cacheKey = `${memoryPrefix}-${ProductVersion}-${GitHash}-${configurationHash}`;
+        const cache = await open();
+        if (!cache) {
+            return undefined;
+        }
+        const res = await cache.match(cacheKey);
+        if (!res) {
+            return undefined;
+        }
+        return res.arrayBuffer();
+    } catch (ex) {
+        return undefined;
+    }
+}
+
+export async function storeMemory(configurationHash: number, memory: ArrayBuffer) {
+    try {
+        const cacheKey = `${memoryPrefix}-${ProductVersion}-${GitHash}-${configurationHash}`;
```

Comment:
Is GitHash actually sufficient? If the source they built the application from has local changes, the git hash will still be that of the most recent commit, perhaps with an -m on the end. Which means if they make more changes and run again, it could use an outdated memory cache, right?

Reply:
There are hashes of .wasm file as well as all DLLs and assets in the `configurationHash` via mono-config.json `assetsHash`

Reply:
`GitHash` is probably not even necessary

Reply:
If we are including the assets hash that's great, the name "configuration hash" made me think it was configuration (like locale and timezone) only

Reply:
I'll rename it and comment more